libraries/adaptive-expressions/src/functionUtils.internal.ts

```diff
@@ -298,48 +298,84 @@ export class InternalFunctionUtils {
+            const list = InternalFunctionUtils.convertToList(instance);
+            if (!list) {
+                error = `${expression.children[0]} is not a collection or structure object to run Foreach`;
             } else {
-                error = `${expression.children[0]} is not a collection or structure object to run foreach`;
-            }
-
-            if (!error) {
-                const stackedMemory = StackedMemory.wrap(state);
                 result = [];
-                for (const item of arr) {
-                    const local: Map<string, any> = new Map<string, any>([[iteratorName, item]]);
-
-                    stackedMemory.push(SimpleObjectMemory.wrap(local));
-                    const { value: r, error: e } = expression.children[2].tryEvaluate(stackedMemory, options);
-                    stackedMemory.pop();
-                    if (e !== undefined) {
-                        return { value: undefined, error: e };
+                InternalFunctionUtils.lambdaEvaluator(expression, state, options, list, (currentItem, r, e) => {
+                    if (e) {
+                        error = e;
+                        return true;
+                    } else {
+                        result.push(r);
+                        return false;
                     }
-                    result.push(r);
-                }
+                });
             }
         }
 
         return { value: result, error };
     }
 
+    /**
+     * Lambda evaluator.
+     * @param expression expression.
+     * @param state memory state.
+     * @param options options.
+     * @param list item list.
+     * @param callback call back. return the should break flag.
+     */
+    public static lambdaEvaluator(expression: Expression, state: MemoryInterface, options: Options, list: unknown[], callback: (currentItem: unknown, result: unknown, error: string) => boolean): void{
+        const iteratorName = (expression.children[1].children[0] as Constant).value as string;
+        const stackedMemory = StackedMemory.wrap(state);
+        for (const item of list) {
+            const currentItem = item;
+            const local: Map<string, any> = new Map<string, any>([[iteratorName, item]]);
+
+            // the local iterator is pushed as one memory layer in the memory stack
+            stackedMemory.push(SimpleObjectMemory.wrap(local));
+            const { value: r, error: e } = expression.children[2].tryEvaluate(stackedMemory, options);
+            stackedMemory.pop();
+
+            const shouldBreak = callback(currentItem, r, e);
+            if (shouldBreak) {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Convert an object into array.
+     * If the instance is array, return itself.
+     * If the instance is object, return {key, value} pair list.
+     * Else return undefined.
+     * @param instance input instance.
+     */
+    public static convertToList(instance: unknown) : unknown[] | undefined {
```

Comment:
Would `Object.entries` work instead?

Reply:
The browser compatibility of `Object.values` and `Object.entries` is not as good as `Object.keys`.
 And since there is no big difference in the complexity of the code, i kept it. Do you think that make sense or `Object.entries ` has covered all our browser usage scenarios

Reply:
Fair enough!